File: model/src/main/kotlin/net/dungeonhub/model/discord_role/DiscordRoleModel.java
package net.dungeonhub.model.discord_role;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;
import me.taubsie.dungeonhub.common.DungeonHubService;
import me.taubsie.dungeonhub.common.entity.model.Model;
import me.taubsie.dungeonhub.common.model.server.DiscordServerModel;

@Getter
@Setter
@AllArgsConstructor
public class DiscordRoleModel implements Model {
    private long id;
    private String nameSchema;
    private boolean verifiedRole;
    private DiscordServerModel discordServerModel;

    public static DiscordRoleModel fromJson(String json) {
        return DungeonHubService.getInstance()
                .getGson()
                .fromJson(json, DiscordRoleModel.class);
    }
}

File: model/src/main/kotlin/net/dungeonhub/model/discord_role/DiscordRoleModel.kt
package net.dungeonhub.model.discord_role

import net.dungeonhub.enums.RoleAction
import net.dungeonhub.model.discord_server.DiscordServerModel
import net.dungeonhub.service.MoshiService
import net.dungeonhub.structure.model.UpdateableModel

class DiscordRoleModel(
    val id: Long,
    val nameSchema: String?,
    val roleAction: RoleAction,
    val discordServerModel: DiscordServerModel
) : UpdateableModel<DiscordRoleUpdateModel, DiscordRoleModel> {
    companion object {
        fun fromJson(json: String): DiscordRoleModel {
            return MoshiService.moshi.adapter(DiscordRoleModel::class.java).fromJson(json)!!
        }
    }

    fun shouldBeAdded(verified: Boolean): Boolean {
        return if (verified) {
            RoleAction.applyWhenVerified.contains(roleAction)
        } else {
            RoleAction.applyWhenUnverified.contains(roleAction)
        }
    }

    fun shouldBeRemoved(verified: Boolean): Boolean {
        return if (verified) {
            RoleAction.removeWhenVerified.contains(roleAction)
        } else {
            RoleAction.removeWhenUnverified.contains(roleAction)
        }
    }

    override fun getUpdateModel(): DiscordRoleUpdateModel {
        return DiscordRoleUpdateModel(null, null)
    }
}